ironfish/src/network/messages/disconnecting.ts

```diff
@@ -0,0 +1,80 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import bufio from 'bufio'
+import { Identity, identityLength } from '../identity'
+import { NetworkMessage, NetworkMessageType } from './networkMessage'
+
+export enum DisconnectingReason {
+  ShuttingDown = 0,
+  Congested = 1,
+}
+
+interface CreateDisconnectingMessageOptions {
+  // Can be null if we're sending the message to an unidentified Peer
+  destinationIdentity: Identity | null
+  // Expects a timestamp with millisecond precision
+  disconnectUntil: number
+  reason: DisconnectingReason
+  sourceIdentity: Identity
+}
+
+export class DisconnectingMessage extends NetworkMessage {
+  readonly destinationIdentity: Identity | null
+  readonly disconnectUntil: number
+  readonly reason: DisconnectingReason
+  readonly sourceIdentity: Identity
+
+  constructor({
+    destinationIdentity,
+    disconnectUntil,
+    reason,
+    sourceIdentity,
+  }: CreateDisconnectingMessageOptions) {
+    super(NetworkMessageType.Disconnecting)
+    this.destinationIdentity = destinationIdentity
+    this.disconnectUntil = disconnectUntil
+    this.reason = reason
+    this.sourceIdentity = sourceIdentity
+  }
+
+  serialize(): Buffer {
+    const bw = bufio.write(this.getSize())
+    // Truncates the timestamp to seconds
+    bw.writeU32(Math.ceil(this.disconnectUntil / 1000))
+    bw.writeU8(this.reason)
+    bw.writeBytes(Buffer.from(this.sourceIdentity, 'base64'))
+    if (this.destinationIdentity) {
+      bw.writeBytes(Buffer.from(this.destinationIdentity, 'base64'))
+    }
+    return bw.render()
+  }
+
+  static deserialize(buffer: Buffer): DisconnectingMessage {
+    const reader = bufio.read(buffer, true)
+    const disconnectUntil = reader.readU32() * 1000
+    const reason = reader.readU8()
+    const sourceIdentity = reader.readBytes(identityLength).toString('base64')
+    let destinationIdentity = null
+    if (reader.left()) {
+      destinationIdentity = reader.readBytes(identityLength).toString('base64')
+    }
+    return new DisconnectingMessage({
+      destinationIdentity,
+      disconnectUntil,
+      reason,
+      sourceIdentity,
+    })
+  }
+
+  getSize(): number {
```

Comment:
Is `size` the first thing serialized on every Serializable object? It does seems a little error prone to calculate the size of the object separately from the serialization logic. If they get out of sync (like we add a field to `serialize` but forget to add it to `getSize`). Is there a way we combine this logic; maybe get the size of the buffer in `NetworkMessage` and use that as the size?

Reply:
We also use the size when allocating the buffer [here](https://github.com/iron-fish/ironfish/pull/1215/files#diff-933783bbd91ba8f941400e2ce015b750ea4359a65af898ab012e535110e4567cR43)

Reply:
The size from `getSize` is just used to allocate the Buffer when serializing the message data. Tests will catch errors if the buffer is too small, but I agree it's a source of potential issues to keep the numbers in sync.

We could potentially do some abstraction around message field specification, where you'd define a message and its field types, and that would match up to the proper reader/writer and size behind the scenes, but that's probably outside the scope of this PR.